src/common/dav/tx.c

```diff
@@ -452,18 +452,21 @@ tx_create_wal_entry(struct ulog_entry_base *e, void *arg,
 int
 lw_tx_begin(dav_obj_t *pop)
 {
-	struct umem_wal_tx	*utx = pop->do_utx;
+	struct umem_wal_tx	*utx = NULL;
 	int			 rc;
 
-	D_ASSERT(utx == NULL);
-	utx = dav_umem_wtx_new(pop);
-	if (utx == NULL)
-		return obj_tx_fail_err(EINVAL, 0);
+	if (pop->do_utx == NULL) {
+		utx = dav_umem_wtx_new(pop);
+		if (utx == NULL)
+			return obj_tx_fail_err(EINVAL, 0);
+	}
 	rc = dav_wal_tx_reserve(pop);
```

Comment:
I think the WAL reserve should be called only once for a umem_wal_tx, maybe we'd combine the dav_umem_wtx_new() & dav_wal_tx_reserve() to single function?

Reply:
The current implementation allows creation of wal entries without a need to call wal_reserve. This helps in the implementation of dav_reserve() where the changes it does to memory can be persisted during the next transaction.

Reply:
Ok, but we need to avoid calling WAL reserve multiple times for a umem_wal_tx, maybe introducing a flag in umem_wal_tx to indicate WAL reserve is called? (or take UINT64_MAX as invalid/uninitialized WAL ID).

Reply:
We call wal_reserve for only two cases:
1. Outer most tx begin
2. Beginning of an atomic operation
The only scenario where wal reserve gets called again within a transaction is when atomic calls are made from within a transaction. Since we cannot support this scenario, I am thinking whether placing an assert in the bmem wrappers would suffice?  Is that ok with you?

Reply:
Sure, then it'll be great to add an assert(utx_id == ULLONG_MAX) in dav_wal_tx_reserve().

Reply:
Done.